csrc/scheduler/normalization.cpp

```diff
@@ -943,200 +978,179 @@ std::shared_ptr<ReductionParams> outerPersistentHeuristic(
-  if (max_threads_in_block % warp_size != 0) {
-    max_threads_in_block += warp_size - max_threads_in_block % warp_size;
-  }
-
   // Compute maximum number of reductions we could do in the same kernel based
   // on persistent buffer size
   const int64_t max_multi_reduction_factor = scheduler_utils::safeDiv(
       scheduler_utils::register_file_size, max_persistent_buffer_size);
 
-  // To get to target threads:
-  // Prioritize
-  // (1) x dim in iter domain
-  // (2) unrolling in iter domain
-  // (3) y in reduction domain
-  // To get target blocks:
-  // Prioritize
-  // (1) x dim in multiple outputs
-  // (2) y dim in multiple reductions - need to flip unrolling to reduction
-  // domain for this
-
-  // Threads for reduction
-  int64_t bdimy = 1;
-  // Threads for output
-  int64_t bdimx = 1;
-
-  // Unroll amount
-  int64_t inner_reduction_unroll_factor = 1;
-  int64_t iter_unroll_factor = 1;
-
-  // If we only use a warp, can we get iter domain unrolling?
-  bdimx = std::min(max_multi_reduction_factor, warp_size);
-  // Round down if it didn't hit a full warp
-  if (bdimx < warp_size) {
-    bdimx = scheduler_utils::lastPow2(bdimx);
+  struct HeuristicParams {
+    // Iteration dim, each CTA covers [bdimx] * [iter_unroll_factor] reductions.
+    // Needs total_iteration_numel / (bdimx * iter_unroll_factor) CTAs.
+    HeuristicParameterWrapper iter_unroll_factor;
+    HeuristicParameterWrapper bdimx;
+    // Reduction dim, each thread do [batches_per_block * redu_unroll_factor]
+    // serial reductions, then do block reductions along [bdimy].
+    // Total_reduction_numel <= bdimy [dynamic] * batches_per_block *
+    // redu_unroll_factor
+    HeuristicParameterWrapper redu_unroll_factor;
+    HeuristicParameterWrapper batches_per_block;
+    HeuristicParameterWrapper bdimy;
+    void verify() {
+      TORCH_INTERNAL_ASSERT(
+          !iter_unroll_factor.isMutable(),
+          "iter_unroll_factor is not finalized.");
+      TORCH_INTERNAL_ASSERT(!bdimx.isMutable(), "bdimx is not finalized.");
+      TORCH_INTERNAL_ASSERT(
+          !redu_unroll_factor.isMutable(),
+          "redu_unroll_factor is not finalized.");
+      TORCH_INTERNAL_ASSERT(
+          !batches_per_block.isMutable(),
+          "batches_per_block is not finalized.");
+      TORCH_INTERNAL_ASSERT(!bdimy.isMutable(), "bdimy is not finalized.");
+    }
+  };
+  HeuristicParams hp;
+
+  // set iter_unroll_factor
+  // This controls vectorized load/store along the iteration dimension.
+  // The kernel calls block reduction [iter_unroll_factor] times.
+  // Test shows performance regression when iter_unroll_factor > 1 due to
+  // the high cost of calling block reduction multiple times per block.
+  // However, if total_iteration_numel is extremely large, set
+  // iter_unroll_factor > 1 can reduce the required number of CTAs. Here 2097152
+  // is chosen as the threshold based on test results on A100.
+  const int64_t large_size = 2097152l;
```

Comment:
Can this be parameterized with device properties such as the number of SMs?

Reply:
no longer being used.